src/BlockTypes/Checkout.php

```diff
@@ -48,63 +50,74 @@ public function register_block_type() {
+			AssetDataRegistry::class
 		);
-		if ( ! empty( $attributes['cartPageId'] ) && ! $data_registry->exists( 'page-' . $attributes['cartPageId'] ) ) {
-			$permalink = get_permalink( $attributes['cartPageId'] );
-			if ( $permalink ) {
-				$data_registry->add( 'page-' . $attributes['cartPageId'], get_permalink( $attributes['cartPageId'] ) );
-			}
-		}
-		if ( ! $data_registry->exists( 'allowedCountries' ) ) {
-			$data_registry->add( 'allowedCountries', WC()->countries->get_allowed_countries() );
-		}
-		if ( ! $data_registry->exists( 'shippingCountries' ) ) {
-			$data_registry->add( 'shippingCountries', WC()->countries->get_shipping_countries() );
-		}
-		if ( ! $data_registry->exists( 'allowedStates' ) ) {
-			$data_registry->add( 'allowedStates', WC()->countries->get_allowed_country_states() );
-		}
-		if ( ! $data_registry->exists( 'shippingStates' ) ) {
-			$data_registry->add( 'shippingStates', WC()->countries->get_shipping_country_states() );
-		}
-		if ( ! $data_registry->exists( 'cartData' ) ) {
-			$data_registry->add( 'cartData', WC()->api->get_endpoint_data( '/wc/store/cart' ) );
-		}
-		if ( ! $data_registry->exists( 'billingData' ) && WC()->customer instanceof \WC_Customer ) {
-			$data_registry->add( 'billingData', WC()->customer->get_billing() );
-		}
-		if ( function_exists( 'get_current_screen' ) ) {
-			$screen = get_current_screen();
-			if ( $screen && $screen->is_block_editor() && ! $data_registry->exists( 'shippingMethodsExist' ) ) {
-				$methods_exist = wc_get_shipping_method_count() > 0;
-				$data_registry->add( 'shippingMethodsExist', $methods_exist );
+
+		$block_data = [
+			'allowedCountries'  => [ WC()->countries, 'get_allowed_countries' ],
+			'shippingCountries' => [ WC()->countries, 'get_shipping_countries' ],
+			'allowedStates'     => [ WC()->countries, 'get_allowed_country_states' ],
+			'shippingStates'    => [ WC()->countries, 'get_shipping_country_states' ],
+		];
+
+		foreach ( $block_data as $key => $callback ) {
+			if ( ! $data_registry->exists( $key ) ) {
+				$data_registry->add( $key, call_user_func( $callback ) );
 			}
 		}
-		if ( is_user_logged_in() && ! $data_registry->exists( 'customerPaymentMethods' ) ) {
-			add_filter( 'woocommerce_payment_methods_list_item', [ $this, 'include_token_id_with_payment_methods' ], 10, 2 );
-			$data_registry->add(
-				'customerPaymentMethods',
-				wc_get_customer_saved_methods_list( get_current_user_id() )
-			);
-			remove_filter( 'woocommerce_payment_methods_list_item', [ $this, 'include_token_id_with_payment_methods' ], 10, 2 );
+
+		$this->hydrate_from_api( $data_registry );
+		$this->hydrate_customer_payment_methods( $data_registry );
+
+		$permalink = ! empty( $attributes['cartPageId'] ) ? get_permalink( $attributes['cartPageId'] ) : false;
+
+		if ( $permalink && ! $data_registry->exists( 'page-' . $attributes['cartPageId'] ) ) {
+			$data_registry->add( 'page-' . $attributes['cartPageId'], get_permalink( $attributes['cartPageId'] ) );
```

Comment:
Does there need to be another call to `get_permalink` here? Could you use `$permalink`?